### ISSUE TITLE
datalab connect is hanging when using gcloud container builds submit with simple Dockerfile 

### ISSUE THREAD

**Issue by eilalan** [googledatalab/datalab#1928]
Hello,
Could you please advice why the following minimal dockerfile is unable to perfrom "datalab connect" command. 

The steps for container creation are the following:
**Running on MAC machine:**
Docker file:

```
FROM gcr.io/cloud-datalab/datalab:latest

RUN apt-get update

ENTRYPOINT [ "/datalab/run.sh" ]
```
**Build container using the command (only Dockerfile in folder)**
`gcloud container builds submit --tag gcr.io/orielresearch-188115/datalab_orig .
`
**Running on google cloud shell:**
create a datalab contianer:
`datalab create --image-name datalab_orig datalab-test
`

**The execution is stuck at:**
Waiting for Datalab to be reachable at http://localhost:8081/ 

**With the following error:**
channel 1: open failed: connect failed: Connection refused
debug1: channel 1: free: direct-tcpip: listening port 8081 for localhost port 8080, connect from 127.0.0.1 port 42885 to 127.0.0.1 port 8081, nchannels 2
debug1: Connection to port 8081 forwarding to localhost port 8080 requested.

Please let me know if you have any idea how to debug / what is wrong / missing with the above process.

Many thanks,
Eila

**Comment by harmon** [comment]
First, I would check if there are any insights in the output from:

```
datalab connect --verbosity debug datalab-test
```

Second, I would SSH into the instance (gcloud compute ssh datalab-test) and check two things:

1. Output from "docker ps" to make sure the datalab docker container is running.
2. Check inside /var/log/startupscript.log (I think it's named that?). That usually points me to any errors in kicking off the docker startup.

**Comment by chmeyers** [comment]
The gcr.io/orielresearch-188115/datalab_orig image is private, right?  This may be a permissions issue with the datalab VM may not be able to successfully pull the image down.  As harmon said above, you should SSH into the VM and see if the container is even running with 'docker ps' and if not, whether the image exists on the machine with 'docker images' and if not, whether you can pull it down on the machine manually.

https://github.com/googledatalab/datalab/issues/1922 has some discussion on this topic.

**Comment by eilalan** [comment]
Hi,

I have tried to use datalab built-in method to launch an updated datalab container- [https://github.com/googledatalab/datalab/blob/master/containers/datalab/Dockerfile-extended-example.in](url)  (with this option, I was able to populate the image in the VM - `docker images` command) and took the following steps to allow permission to docker, datalab on gcloud. Everything is being run from the cloud shell (no local machine is involved)

1. Login to docker by `docker login`
2. Login to google by `gcloud auth login`
3. git clone the datalab repo. Update datalab-extended.sh and the Dockerfile-extended-example.in
5. Dockerfile-extended-example.in was udpated with gcsfuse installation:
```
RUN apt-get update
RUN apt-get install -y lsb-release
RUN export GCSFUSE_REPO=gcsfuse-`lsb_release -c -s`
RUN bash -c 'echo "deb http://packages.cloud.google.com/apt gcsfuse-`lsb_release -c -s` main" | tee /etc/apt/sources.list.d/gcsfuse.list'
RUN curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
RUN apt-get update
RUN apt-get install -y gcsfuse

COPY XXXXXXX.json /datalab/gcloud-creds.json
```

5. and the run-extended.sh
```
if [ -f custom-packages.txt ] || [ -f Dockerfile-extended.in ];
then
    if [ -f custom-packages.txt ];
    then
        # First create a new Docker file called Dockerfile-custom-packages. Start with the standard image
        # TODO: at some point the local Datalab container will be tagged 'latest' rather than 'local'
        # and the line below should change.
        echo 'FROM gcr.io/cloud-datalab/datalab:local' > Dockerfile-custom-packages

        # Add the script with a list of custom packages to the Dockerfile
        echo 'ADD custom-packages.txt /datalab/custom-packages.txt' >> Dockerfile-custom-packages
        echo 'RUN pip install -r /datalab/custom-packages.txt' >> Dockerfile-custom-packages

        DOCKERFILE=Dockerfile-custom-packages
        DOCKERIMAGE=datalab-custom-packages

    elif [ -f Dockerfile-extended.in ];
    then
        DOCKERFILE=Dockerfile-extended.in
        DOCKERIMAGE=datalab-extended
    fi

    # Build the customized docker image derived from the standard datalab image
    docker build ${DOCKER_BUILD_ARGS} -t $DOCKERIMAGE -f $DOCKERFILE .
    # ADDITION: Here's where your bucket filesystem will be mounted:
    mkdir -p /content/datalab/bucket-share
    # ADDITION: This mounts the bucket using your credentials built into the Dockerfile:
    gcsfuse --key-file /datalab/orielresearch-1e395b7a563b.json archs4 /content/datalab/bucket-share

fi

```

6. Create.py docker run was added: +   --cap-add SYS_ADMIN --device=/dev/fuse \

**The connection is keep trying with no success. I have checked the commands that were suggested and following are the results:**
1. The gcr.io/orielresearch-188115/datalab_orig image is private, right? _yes. it has the GCS credential in it. If making it public is the only solution, how can I decouple the credentials from the image?_
2. SSH into the VM and see if the container is even running with 'docker ps' . _The only ps is (which is not my datalab-test)_
```
CONTAINER ID        IMAGE                                       COMMAND                  CREATED             STATUS
              PORTS               NAMES
4ab4898cb839        gcr.io/google_containers/fluentd-gcp:1.18   "/bin/sh -c '/usr/..."   27 minutes ago      Up 27 
minutes                           logger
```
3. whether the image exists on the machine with 'docker images' . _Yes, it exists (previous version with the manually created image was not generating the image)_
```
REPOSITORY                             TAG                 IMAGE ID            CREATED             SIZE
gcr.io/cloud-datalab/datalab           latest              3d77c19f6fba        2 weeks ago         3.34 GB
gcr.io/google_containers/fluentd-gcp   1.18                e0be8d052951        23 months ago       411 MB
```
4. /var/log/startupscript.log
`This was generated when I used the --image-name. I could not find this log file when running with the docker-extended-sample.in version`

5. Running datalab.py create --vebosity debug datalab-test. _The following command is firing the permission denied:_
```
DEBUG: Executing command: ['/usr/bin/ssh', '-t', '-i', '/home/eila/.ssh/google_compute_engine', '-o', 'CheckHostIP=no', '-o', 'HostKeyAlias=compute.116593348
8214550037', '-o', 'IdentitiesOnly=yes', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/home/eila/.ssh/google_compute_known_hosts', u'-o', u'Lo
gLevel=error', u'-4', u'-N', u'-L', u'localhost:8081:localhost:8080', u'datalab@35.194.74.81']
Permission denied (publickey).
```

**My questions are:**
1. At step 5 (run-extended.sh), is it correct to add the mount after the build of the docker?
2. Could you please explain the google_compute_engine -o permission denied (publickey) 
3. Where is the permission issue that you are referring to? I don't know what permission issue it is facing? Is is for downloading the image? Now that the image is available on the vm, what will be the next permission issue? accessing the image for installation?
4. Where can I find the log file such as startupscript.log
5. Any other hint?

Many thanks,
Eila

**Comment by harmon** [comment]
Can you `gcloud compute ssh <my_instance>` into the host, and try the following?

```
$ docker-credential-gcr configure-docker
$ docker pull gcr.io/orielresearch-188115/datalab_orig
```

If that doesn't work, it's probably an issue with the service account tied to the instance. You'll have to give it a few permissions, notably: "Storage > Storage Object Viewer". But I would just give it "Project > Editor" permissions to start, just to make sure it can access stuff. Then you can bump it's permissions back down.

Also, you said /var/log/startupscript.log was created when you did "datalab create --image-name gcr.io/orielresearch-188115/datalab_orig <my_instance>". Did you look through that file for clues as to what went wrong? Mine had logs of docker errors when trying to pull a private image until I got permissions set correctly. So look through there and see if there are any telling errors.

**Comment by harmon** [comment]
Eila: The following lines need to be run IN the docker container when the datalab.service starts on your remove host. You've place those lines in a build script that runs on YOUR computer when you are initially building the image:

```
# ADDITION: Here's where your bucket filesystem will be mounted:
    mkdir -p /content/datalab/bucket-share
    # ADDITION: This mounts the bucket using your credentials built into the Dockerfile:
    gcsfuse --key-file /datalab/orielresearch-1e395b7a563b.json archs4 /content/datalab/bucket-share
```

These need to run IN the docker container AFTER you deploy it to the Google Compute host. That's why I suggested putting in into a new "ENTRYPOINT ["/datalab/run-custom.sh"]" in the other Github Issue. That's the script that runs when systemctl runs the datalab.service.

**Comment by eilalan** [comment]
Hi Harmon,
Thank you for your help. With all my frustration to figure this out, your comments are very helpful and direct me to the relevant things to learn and try.

One issue that I was able to explain is the docker ps empy output. This occurred becuase I was using --image-name datalab-basic which is translated to gcr.io/orielresearch-188115/datalab-basic while the images are saved on us.gcr.io. Once this was explicitly written --image-name=us.gcr.io/..., I was able to get the docker ps to return the error message and
docker-credential-gcr configure-docker was executed successfully when running datalab create...

Following all the trials and their outputs:
```
$ docker-credential-gcr configure-docker
/home/eila/.docker/config.json configured to use this credential helper for GCR registries
$docker pull us.gcr.io/orielresearch-188115/datalab-basic
Using default tag: latest
Error response from daemon: repository us.gcr.io/orielresearch-188115/datalab-basic not found: does not exist 
or no pull access
```
This didnt work by itself, so I added the following for the pull command: 
```
$docker-credential-gcr gcr-login
Please visit the following URL and complete the authorization dialog:
https://accounts.google.com/o/oauth2/v2/auth?access_type=offline&client_id=99426463878-o7n0bshgue20tdpm25q4at0
vs2mr4utq.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&response_type=code&scope
=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform&state=state
Authorization code:
blablabal
```

Pull command - which returns the same error as seen below at the startup log:
```
$docker pull us.gcr.io/orielresearch-188115/datalab-basic
Using default tag: latest
Error response from daemon: repository us.gcr.io/orielresearch-188115/datalab-basic not found: does not exist 
or no pull access

```
Just to be on the safe side, the docker image is available on the registry:
```
$ gcloud container images list-tags us.gcr.io/orielresearch-188115/datalab-basic
DIGEST        TAGS    TIMESTAMP
dcfe2ea2acb3  latest  2018-01-29T10:17:45
```

The logs from /var/log/startupscript.log:
```
$docker ps
CONTAINER ID        IMAGE                                       COMMAND                  CREATED             STATUS              PORTS               NAMES
22a8ce90dea6        gcr.io/google_containers/fluentd-gcp:1.18   "/bin/sh -c '/usr/..."   37 minutes ago      Up 37 minutes                           logger

```
The log file with grep for error:
```
$ cat /var/log/startupscript.log | grep -i err
Jan 30 18:15:39 basic4 startup-script[814]: INFO startup-script: Error response from daemon: repository us.gcr.io/orielresearch-188115/datalab-basic not f
ound: does not exist or no pull access
Jan 30 18:15:39 basic4 startup-script[814]: INFO startup-script:        missing codepage or helper program, or other error
Jan 30 18:16:17 basic4 startup-script[814]: INFO startup-script: docker: Error response from daemon: repository us.gcr.io/orielresearch-188115/datalab-bas
ic not found: does not exist or no pull access.
Jan 30 18:16:33 basic4 startup-script[814]: INFO startup-script: docker: Error response from daemon: Get https://us.gcr.io/v2/orielresearch-188115/datalab
-basic/manifests/latest: Get https://us.gcr.io/v2/token?scope=repository%3Aorielresearch-188115%2Fdatalab-basic%3Apull&service=us.gcr.io: net/http: reques
t canceled while waiting for connection (Client.Timeout exceeded while awaiting headers).
```

More complete log:
```
$cat /var/log/startupscript.log
Jan 30 18:15:37 basic4 systemd[1]: Starting Google Compute Engine Startup Scripts...
Jan 30 18:15:37 basic4 startup-script[814]: INFO Starting startup scripts.
Jan 30 18:15:37 basic4 startup-script[814]: INFO Found startup-script in metadata.
Jan 30 18:15:37 basic4 startup-script[814]: INFO startup-script: Getting Docker credentials
Jan 30 18:15:38 basic4 startup-script[814]: INFO startup-script: /home/datalab/.docker/config.json configured to use this credential helper for GCR registries
Jan 30 18:15:38 basic4 startup-script[814]: INFO startup-script: Pulling latest image: us.gcr.io/orielresearch-188115/datalab-basic
Jan 30 18:15:38 basic4 startup-script[814]: INFO startup-script: Using default tag: latest
Jan 30 18:15:39 basic4 startup-script[814]: INFO startup-script: Error response from daemon: repository us.gcr.io/orielresearch-188115/datalab-basic not found: does not exist or no pull access
Jan 30 18:15:39 basic4 startup-script[814]: INFO startup-script: Trying to mount the persistent disk
Jan 30 18:15:39 basic4 startup-script[814]: INFO startup-script: mount: wrong fs type, bad option, bad superblock on /dev/sdb,
Jan 30 18:15:39 basic4 startup-script[814]: INFO startup-script:        missing codepage or helper program, or other error
Jan 30 18:15:39 basic4 startup-script[814]: INFO startup-script:        In some cases useful info is found in syslog - try
Jan 30 18:15:39 basic4 startup-script[814]: INFO startup-script:        dmesg | tail or so.
Jan 30 18:15:39 basic4 startup-script[814]: INFO startup-script: Checking if the persistent disk needs to be formatted
Jan 30 18:15:39 basic4 startup-script[814]: INFO startup-script: Formatting the persistent disk
Jan 30 18:15:39 basic4 startup-script[814]: INFO startup-script: mke2fs 1.43.6 (29-Aug-2017)
Jan 30 18:15:40 basic4 startup-script[814]: [153B blob data]
Jan 30 18:15:40 basic4 startup-script[814]: INFO startup-script: Creating filesystem with 52428800 4k blocks and 13107200 inodes
Jan 30 18:15:40 basic4 startup-script[814]: INFO startup-script: Filesystem UUID: 4bc2a9c4-2de7-408b-ac80-d2d910804b48
Jan 30 18:15:40 basic4 startup-script[814]: INFO startup-script: Superblock backups stored on blocks:
Jan 30 18:15:40 basic4 startup-script[814]: INFO startup-script:         32768, 98304, 163840, 229376, 294912, 819200, 884736, 1605632, 2654208,
Jan 30 18:15:40 basic4 startup-script[814]: INFO startup-script:         4096000, 7962624, 11239424, 20480000, 23887872
Jan 30 18:15:40 basic4 startup-script[814]: [86B blob data]
Jan 30 18:16:03 basic4 startup-script[814]: [497B blob data]
Jan 30 18:16:14 basic4 startup-script[814]: INFO startup-script: Creating journal (262144 blocks): done
Jan 30 18:16:17 basic4 startup-script[814]: [156B blob data]
Jan 30 18:16:17 basic4 startup-script[814]: INFO startup-script: Creating the datalab directory
Jan 30 18:16:17 basic4 startup-script[814]: INFO startup-script: Cloning the repo datalab-notebooks
Jan 30 18:16:17 basic4 startup-script[814]: INFO startup-script: Unable to find image 'us.gcr.io/orielresearch-188115/datalab-basic:latest' locally
```

and for the datalab start logs, it was not able to find the datalab without the image pull:
```
$ sudo journalctl -u datalab
Jan 30 18:17:06 basic4 systemd[1]: Starting datalab docker container...
Jan 30 18:17:07 basic4 docker-credential-gcr[1168]: /home/datalab/.docker/config.json configured to use this credential helper for GCR registries
Jan 30 18:17:07 basic4 systemd[1]: Started datalab docker container.
Jan 30 18:17:07 basic4 docker[1178]: Unable to find image 'us.gcr.io/orielresearch-188115/datalab-basic:latest' locally
Jan 30 18:17:07 basic4 docker[1178]: /usr/bin/docker: Error response from daemon: repository us.gcr.io/orielresearch-188115/datalab-basic not found: does 
Jan 30 18:17:07 basic4 docker[1178]: See '/usr/bin/docker run --help'.
Jan 30 18:17:07 basic4 systemd[1]: datalab.service: Main process exited, code=exited, status=125/n/a
Jan 30 18:17:07 basic4 systemd[1]: datalab.service: Unit entered failed state.
Jan 30 18:17:07 basic4 systemd[1]: datalab.service: Failed with result 'exit-code'.
Jan 30 18:17:08 basic4 systemd[1]: datalab.service: Service hold-off time over, scheduling restart.
Jan 30 18:17:08 basic4 systemd[1]: Stopped datalab docker container.
Jan 30 18:17:08 basic4 systemd[1]: Starting datalab docker container...
Jan 30 18:17:08 basic4 docker-credential-gcr[1229]: /home/datalab/.docker/config.json configured to use this credential helper for GCR registries
Jan 30 18:17:08 basic4 systemd[1]: Started datalab docker container.
Jan 30 18:17:08 basic4 docker[1237]: Unable to find image 'us.gcr.io/orielresearch-188115/datalab-basic:latest' locally
Jan 30 18:17:12 basic4 docker[1237]: /usr/bin/docker: Error response from daemon: repository us.gcr.io/orielresearch-188115/datalab-basic not found: does 
Jan 30 18:17:12 basic4 docker[1237]: See '/usr/bin/docker run --help'.
Jan 30 18:17:12 basic4 systemd[1]: datalab.service: Main process exited, code=exited, status=125/n/a
Jan 30 18:17:12 basic4 systemd[1]: datalab.service: Unit entered failed state.
Jan 30 18:17:12 basic4 systemd[1]: datalab.service: Failed with result 'exit-code'.
Jan 30 18:17:14 basic4 systemd[1]: datalab.service: Service hold-off time over, scheduling restart.
Jan 30 18:17:14 basic4 systemd[1]: Stopped datalab docker container.
Jan 30 18:17:14 basic4 systemd[1]: Starting datalab docker container...
```

So, it seems that this is an image pull permission issue / access to private registry. I have tried to add the owner role to the VM service account and it didn't change anything.
I can see the VM service for the VM: XXXX-compute@developer.gserviceaccount.com on the IM&Admin=>service accounts. When I use it to add permissions, I can add members..this confuses me... where is it possible to assign a project editor permission to the service?
I don't see this service on the IAM tab. Should I see it in order to be able update its permission? 
Many thanks,
Eila

**Comment by chmeyers** [comment]
eilalan: See https://cloud.google.com/container-registry/docs/access-control

TL;DR: You need to give it access to the Google Cloud Storage bucket that contains the registry.  After doing that you have to recursively apply those permissions to existing objects as well.

**Comment by harmon** [comment]
@eilalan:

@chmeyers advice is spot on!

Yeah, sounds like a service account permission issue. You'll need to update the Role assigned to the service account under the "IAM & admin > IAM" section, NOT "IAM & admin > Service Accounts" section.

The Service Accounts "Permissions" defines who has access to that Service Account, whereas the "IAM" section defines what permissions the Service Account itself has, ie. what services IT can access.

As a start, grant the service account the "Project > Editor" Role, or just even "Storage > Storage Object View" role, and that *should* be enough for it to pull down the image.

Sounds like you are almost there!

**Comment by eilalan** [comment]
Thank you for the quick response.
I understand the the container registry has a bucket assigned to it and that the service account need a permission to access it. Who is this service account that need to have access it? is it XXXX-compute@developer.gserviceaccount.com ? If yes, it is not on the IAM section. anyone from the below?
```

@container-engine-robot.iam.gserviceaccount.com | Container Engine Service Agent |  
@containerregistry.iam.gserviceaccount.com | Multiple |  
@dataflow-service-producer-prod.iam.gserviceaccount.com | Cloud Dataflow Service Agent |  
@genomics-api.google.com.iam.gserviceaccount.com | Genomics Service Agent |  
@sourcerepo-service-accounts.iam.gserviceaccount.com | Cloud Source Repositories Service Agent |   
@cloudbuild.gserviceaccount.com Cloud Container Builder
@cloudservices.gserviceaccount.com
```

**Comment by harmon** [comment]
You said "XXXX-compute@developer.gserviceaccount.com" was the service account of the datalab VM that got spun up, correct? That means it's the "Compute Engine default service account". To add it to IAM, go to "IAM & admin > IAM", then click "ADD", and then select that from the list, and assign it one of the roles I discussed above.

**Comment by eilalan** [comment]
Thank you all. Basic docker is working. I am switching back to the mounting task. Many thanks!
